### PR TITLE
[everflow] Improve the method of monitor port parsing

### PR DESCRIPTION
**Pull request by steven-guo-ec**
Signed-off-by: Steven Guo steven_guo@edge-core.com

### Description of PR

Summary: Improve the method of monitor port parsing
Fixes #2302
### Type of change

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [X] Test case(new/improvement)

#### What is the motivation for this PR?
The modification for parsing monitor in #2302 used hard-coded method to get the monitor port and imply that the "SRC Port" and "Direction" are not configured.
#### How did you do it?
We parsed the output of mirror_session CLI and create objects to store values of each fields in SPAN and ERSPAN, then return the monitor port from the objects. 
#### How did you verify/test it?
We ran the test on t1 topology and all the test passed:
```
============================= test session starts ==============================
platform linux2 -- Python 2.7.12, pytest-4.6.5, py-1.9.0, pluggy-0.13.1 -- /usr/bin/python
cachedir: .pytest_cache
metadata: {'Python': '2.7.12', 'Platform': 'Linux-4.4.0-87-generic-x86_64-with-Ubuntu-16.04-xenial', 'Packages': {'py': '1.9.0', 'pytest': '4.6.5', 'pluggy': '0.13.1'}, 'Plugins': {u'repeat': u'0.8.0', u'xdist': u'1.28.0', u'ansible': u'2.2.2', u'forked': u'1.3.0', u'html': u'1.22.1', u'metadata': u'1.10.0'}}
ansible: 2.8.12
rootdir: /var/johnar/sonic-mgmt/tests, inifile: pytest.ini
plugins: ansible-2.2.2, metadata-1.10.0, repeat-0.8.0, forked-1.3.0, html-1.22.1, xdist-1.28.0
collecting ... collected 20 items

everflow/test_everflow_ipv6.py::TestIngressEverflowIPv6::test_src_ipv6_mirroring[cli]
everflow/test_everflow_ipv6.py::TestIngressEverflowIPv6::test_dst_ipv6_mirroring[cli]
everflow/test_everflow_ipv6.py::TestIngressEverflowIPv6::test_next_header_mirroring[cli]
everflow/test_everflow_ipv6.py::TestIngressEverflowIPv6::test_l4_src_port_mirroring[cli]
everflow/test_everflow_ipv6.py::TestIngressEverflowIPv6::test_l4_dst_port_mirroring[cli]
everflow/test_everflow_ipv6.py::TestIngressEverflowIPv6::test_l4_src_port_range_mirroring[cli]
everflow/test_everflow_ipv6.py::TestIngressEverflowIPv6::test_l4_dst_port_range_mirroring[cli]
everflow/test_everflow_ipv6.py::TestIngressEverflowIPv6::test_tcp_flags_mirroring[cli]
everflow/test_everflow_ipv6.py::TestIngressEverflowIPv6::test_dscp_mirroring[cli]
everflow/test_everflow_ipv6.py::TestIngressEverflowIPv6::test_l4_range_mirroring[cli]
everflow/test_everflow_ipv6.py::TestIngressEverflowIPv6::test_tcp_response_mirroring[cli]
everflow/test_everflow_ipv6.py::TestIngressEverflowIPv6::test_tcp_application_mirroring[cli]
everflow/test_everflow_ipv6.py::TestIngressEverflowIPv6::test_udp_application_mirroring[cli]
everflow/test_everflow_ipv6.py::TestIngressEverflowIPv6::test_any_protocol[cli]
everflow/test_everflow_ipv6.py::TestIngressEverflowIPv6::test_any_transport_protocol[cli]
everflow/test_everflow_ipv6.py::TestIngressEverflowIPv6::test_invalid_tcp_rule[cli]
everflow/test_everflow_ipv6.py::TestIngressEverflowIPv6::test_source_subnet[cli]
everflow/test_everflow_ipv6.py::TestIngressEverflowIPv6::test_dest_subnet[cli]
everflow/test_everflow_ipv6.py::TestIngressEverflowIPv6::test_both_subnets[cli]
everflow/test_everflow_ipv6.py::TestIngressEverflowIPv6::test_fuzzy_subnets[cli]

========================= 20 passed in 167.20 seconds ==========================
```

### CODE REVIEW

File: tests/everflow/everflow_test_utilities.py [hunk 1]
```diff
@@ -602,20 +602,67 @@ def _get_random_src_port(self, setup):
...
+            if ln.strip() == "":
+                break
+
+            index = 0
+            data = {}
+            for s in separators:
+                end = index + len(s)
+                name = header[index:end].strip()
+                value = ln[index:end].strip()
+                index = index + len(s)+ 2
```
Comment (daall) [inline tests/everflow/everflow_test_utilities.py:659]:
```suggestion
                index = header.index(s)
                end = index + len(s)
                name = header[index:end].strip()
                value = ln[index:end].strip()
```

Would it be more robust to use the `index` method that python provides? It seems unlikely that the spacing between columns will change, but if it did it might result in some really odd behavior.

File: tests/everflow/everflow_test_utilities.py [hunk 2]
```diff
@@ -602,20 +602,67 @@ def _get_random_src_port(self, setup):
...
+        pytest_assert(mirror_session["session_name"] in mirror_output["stdout"],
+                      "Test mirror session {} not found".format(mirror_session["session_name"]))
+
+        sessions = {}
+        lns = mirror_output["stdout"].split("\n")
+        while True:
+            obj, lns = self._objectize(lns)
+            if obj == None:
+                break
+            sessions[obj["name"]] = obj
```
Comment (daall) [inline tests/everflow/everflow_test_utilities.py:616]: This part was really hard to follow, what I would recommend is:

1) Move this loop into a helper called `parse_mirror_session_running_config` or something similar so 1) it's easier for others to re-use (`objectize` is helpful but most people will want the functionality provided by this loop as well if they are parsing mirror session data) and 2) it's more clear what this step is accomplishing in the context of `get_monitor_port`.
2) Re-name `objectize` to `parse_mirror_session_group` or something similar. My biggest confusion looking over this code is that I wasn't sure exactly what part of the output `objectize` was operating on/what data it was returning.

Alternatively, you could provide a helper method like `objectize` that accepts an argument for the type of mirror session you are looking for (ERSPAN or SPAN, at this point), and then it would be even more clear what data you're getting back when you call the method.

**Review by daall — CHANGES_REQUESTED** [review]

**Comment by daall** [comment]
> This pull request **introduces 1 alert** when merging [839c0d1](https://github.com/Azure/sonic-mgmt/commit/839c0d140acb41150e075fbcad470814576f353b) into [d062c61](https://github.com/Azure/sonic-mgmt/commit/d062c618b79d42c738c24ae90e9a68f4e9fa517d) - [view on LGTM.com](https://lgtm.com/projects/g/Azure/sonic-mgmt/rev/pr-71b4a9085b4b1c664ff75e018995ac865ff8025b)
> 
> **new alerts:**
> 
> * 1 for Testing equality to None

Please address this warning as well. Thanks!

File: tests/everflow/everflow_test_utilities.py (hunk 1 above)
Comment (steven-guo-ec) [reply tests/everflow/everflow_test_utilities.py:659]: The "s" is a separator in separators list, so `index = header.index(s)` might not correct. Also, `index = separators.index(s)` might be wrong too because each "s" might have the same content, so we coded this way to separate each session field.

File: tests/everflow/everflow_test_utilities.py [hunk 3]
```diff
@@ -602,20 +602,67 @@ def _get_random_src_port(self, setup):
 
     def _get_monitor_port(self, setup, mirror_session, duthost):
         mirror_output = duthost.command("show mirror_session")
-        logging.info("Running mirror session configuration:\n%s", mirror_output["stdout"])
+        logging.info("mirror session configuration: %s", mirror_output["stdout"])
```
[the remark this answers is not in the document]
Comment (steven-guo-ec) [reply tests/everflow/everflow_test_utilities.py]: This is a mistake, has been fixed.

File: tests/everflow/everflow_test_utilities.py [hunk 4]
```diff
@@ -602,20 +602,67 @@ def _get_random_src_port(self, setup):
...
 
-        matching_session = list(filter(lambda line: line.startswith(mirror_session["session_name"]),
-                                       mirror_output["stdout_lines"]))
-        pytest_assert(matching_session, "Test mirror session {} not found".format(mirror_session["session_name"]))
-        logging.info("Found mirror session:\n%s", matching_session[0])
+        pytest_assert(mirror_session["session_name"] in mirror_output["stdout"],
+                      "Test mirror session {} not found".format(mirror_session["session_name"]))
+
+        sessions = {}
+        lns = mirror_output["stdout"].split("\n")
```
[the remark this answers is not in the document]
Comment (steven-guo-ec) [reply tests/everflow/everflow_test_utilities.py]: Has been fixed.

File: tests/everflow/everflow_test_utilities.py [hunk 5]
```diff
@@ -602,20 +602,67 @@ def _get_random_src_port(self, setup):
...
+        session = [x for x in sessions["ERSPAN Sessions"]["data"] if x["Name"] == mirror_session["session_name"]]
+        pytest_assert(0 < len(session))
+
+        monitor_port = session[0]["Monitor Port"]
 
-        monitor_port = matching_session[0].split()[-1]
         pytest_assert(monitor_port in setup["port_index_map"],
                       "Invalid monitor port:\n{}".format(mirror_output["stdout"]))
-        logging.info("Selected monitor port %s (index=%s)", monitor_port, setup["port_index_map"][monitor_port])
+        logging.info("selected monitor interface %s (port=%s)", monitor_port, setup["port_index_map"][monitor_port])
```
[the remark this answers is not in the document]
Comment (steven-guo-ec) [reply tests/everflow/everflow_test_utilities.py]: Has been fixed.

File: tests/everflow/everflow_test_utilities.py [hunk 6]
```diff
@@ -602,20 +602,67 @@ def _get_random_src_port(self, setup):
...
+        table_name = lns[0]
+        separator_line = lns[2]
+        header = lns[1]
+
+        obj = {
+            "name": table_name,
+            "data": []
+        }
+
+        separators = separator_line.split("  ")
```
[the remark this answers is not in the document]
Comment (steven-guo-ec) [reply tests/everflow/everflow_test_utilities.py]: Has been fixed.

File: tests/everflow/everflow_test_utilities.py [hunk 7]
```diff
@@ -604,18 +604,70 @@ def _get_monitor_port(self, setup, mirror_session, duthost):
...
+            if ln.strip() == "":
+                break
+
+            index = 0
+            data = {}
+            for s in separators:
+                end = index + len(s)
+                name = header[index:end].strip()
+                value = ln[index:end].strip()
+                index = index + len(s)+ 2
```
Comment (daall) [inline tests/everflow/everflow_test_utilities.py:664]: Minor nitpick: could you add a space here?
```suggestion
                index = index + len(s) + 2
```

**Review by daall — APPROVED** [review]
LGTM w/ one small suggestion

File: tests/everflow/everflow_test_utilities.py (hunk 7 above)
Comment (steven-guo-ec) [reply tests/everflow/everflow_test_utilities.py]: Sure, thanks. Has been fixed.

**Comment by daall** [comment]
retest vsimage please

**Comment by daall** [comment]
retest vsimage please
